packages/google-pay-integration/src/gateways/google-pay-gateway.ts

```diff
@@ -0,0 +1,210 @@
+import { round } from 'lodash';
+
+import {
+    AddressRequestBody,
+    BillingAddressRequestBody,
+    guard,
+    InvalidArgumentError,
+    NotInitializedError,
+    NotInitializedErrorType,
+    PaymentIntegrationService,
+    PaymentMethod,
+} from '@bigcommerce/checkout-sdk/payment-integration-api';
+
+import isGooglePayCardNetworkKey from '../guards/is-google-pay-card-network-key';
+import {
+    GooglePayCardDataResponse,
+    GooglePayCardNetwork,
+    GooglePayCardParameters,
+    GooglePayFullBillingAddress,
+    GooglePayGatewayParameters,
+    GooglePayInitializationData,
+    GooglePayMerchantInfo,
+    GooglePaySetExternalCheckoutData,
+    GooglePayTransactionInfo,
+} from '../types';
+
+export default class GooglePayGateway {
+    private _getPaymentMethodFn?: () => PaymentMethod<GooglePayInitializationData>;
+
+    constructor(
+        private _gatewayIdentifier: string,
+        private _paymentIntegrationService: PaymentIntegrationService,
+    ) {}
+
+    mapToShippingAddressRequestBody({
+        shippingAddress,
+    }: GooglePayCardDataResponse): AddressRequestBody | undefined {
+        if (!shippingAddress) {
+            return;
+        }
+
+        const { company = '', phone = '' } =
+            this._paymentIntegrationService.getState().getShippingAddress() || {};
+
+        return this._mapToAddressRequestBody(shippingAddress, company, phone);
+    }
+
+    mapToBillingAddressRequestBody(
+        response: GooglePayCardDataResponse,
+    ): BillingAddressRequestBody | undefined {
+        const { billingAddress } = response.paymentMethodData.info;
+
+        if (!billingAddress) {
+            return;
+        }
+
+        const {
+            company = '',
+            phone = '',
+            email = response.email,
+        } = this._paymentIntegrationService.getState().getBillingAddress() || {};
+
+        return {
+            ...this._mapToAddressRequestBody(billingAddress, company, phone),
+            email,
+        };
+    }
+
+    mapToExternalCheckoutData(
+        response: GooglePayCardDataResponse,
+    ): Promise<GooglePaySetExternalCheckoutData> {
+        const {
+            paymentMethodData: {
+                tokenizationData: { token: nonce },
+                info: { cardNetwork: type, cardDetails: number },
+            },
+        } = response;
+
+        return Promise.resolve({
+            nonce,
+            card_information: { type, number },
+        });
+    }
+
+    getMerchantInfo(): GooglePayMerchantInfo {
+        const {
+            googleMerchantName: merchantName,
+            googleMerchantId: merchantId,
+            platformToken: authJwt,
+        } = this.getGooglePayInitializationData();
+
+        return { merchantName, merchantId, authJwt };
+    }
+
+    getTransactionInfo(): GooglePayTransactionInfo {
+        const { getCheckoutOrThrow, getCartOrThrow } = this._paymentIntegrationService.getState();
```

Comment:
```suggestion
        const { getCheckoutOrThrow, getCartOrThrow: { currency } } = this._paymentIntegrationService.getState();
```

That way you can avoid using getCartOrThrow().currency below.

Reply:
That's not possible because `getCartOrThow` is of type `() => Cart`. So, we need to first call the function to access `currency`. Otherwise, we get the following error: _Property 'currency' does not exist on type '() => Cart'._